README.md
# cryptohash-sha1  [![Build Status](https://travis-ci.org/hvr/cryptohash-sha1.svg?branch=master)](https://travis-ci.org/hvr/cryptohash-sha1)

A practical incremental and one-pass, pure API to the [SHA-1 hash
algorithm](https://en.wikipedia.org/wiki/SHA-1) (including
[HMAC](https://en.wikipedia.org/wiki/HMAC) support) with performance
close to the fastest implementations available in other languages.


changelog.md
## 0.11.101.0

 - Add `Eq` instance for `Ctx`
 - Add `start` and `startlazy` producing `Ctx`
 - Remove ineffective RULES
 - Declare `Crypto.Hash.MD5` module `-XTrustworthy`
 - Convert to `CApiFFI`
 - Added `...AndLength` variants of hashing functions:

      - `finalizeAndLength`
      - `hashlazyAndLength`
      - `hmaclazyAndLength`

 - Minor optimizations in `hmac` and `hash`

## 0.11.100.1

 - Use `__builtin_bswap{32,64}` only with GCC >= 4.3

## 0.11.100.0

 - new `hmac` and `hmaclazy` functions providing HMAC-SHA1
   computation conforming to RFC2104 and RFC2202

## 0.11.7.2

 - switch to 'safe' FFI for calls where overhead becomes negligible
 - removed inline assembly in favour of portable C constructs
 - fix 32bit length overflow bug in `hash` function
 - fix inaccurate context-size
 - add context-size verification to incremental API operations
 - fix unaligned memory-accesses

## 0.11.7.1

 - first version forked off `cryptohash-0.11.7` release


Tests/KAT.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ViewPatterns #-}
import Data.Char
import Data.Bits
import Data.Word
import Data.ByteString (ByteString)
import Data.Foldable (foldl')
import Data.Monoid (mconcat)
import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as BC
import qualified Crypto.Hash.SHA1 as SHA1

import Test.Tasty
import Test.Tasty.QuickCheck
import Test.Tasty.HUnit

v0,v1,v2 :: ByteString
v0 = ""
v1 = "The quick brown fox jumps over the lazy dog"
v2 = "The quick brown fox jumps over the lazy cog"
vectors = [ v0, v1, v2 ]

instance Arbitrary ByteString where
    arbitrary = B.pack `fmap` arbitrary

data HashFct = HashFct
    { fctHash   :: (B.ByteString -> B.ByteString)
    , fctInc    :: ([B.ByteString] -> B.ByteString) }

hashinc i u f = f . foldl u i

sha1Hash = HashFct { fctHash = SHA1.hash, fctInc = hashinc SHA1.init SHA1.update SHA1.finalize }


results :: [ (String, HashFct, [String]) ]
results = [
    ("SHA1", sha1Hash, [
        "da39a3ee5e6b4b0d3255bfef95601890afd80709",
        "2fd4e1c67a2d28fced849ee1bb76e7391b93eb12",
        "de9f2c7fd25e1b3afad3e85a0bd17d9b100db4b3" ])
    ]

hexalise s = concatMap (\c -> [ hex $ c `div` 16, hex $ c `mod` 16 ]) s
        where hex i
                | i >= 0 && i <= 9   = fromIntegral (ord '0') + i
                | i >= 10 && i <= 15 = fromIntegral (ord 'a') + i - 10
                | otherwise          = 0

hexaliseB :: B.ByteString -> B.ByteString
hexaliseB = B.pack . hexalise . B.unpack

splitB :: Int -> ByteString -> [ByteString]
splitB l b =
    if B.length b > l
        then
            let (b1, b2) = B.splitAt l b in
            b1 : splitB l b2
        else
            [ b ]

showHash :: B.ByteString -> String
showHash = map (toEnum.fromEnum) . hexalise . B.unpack

runhash hash v = showHash $ (fctHash hash) $ v
runhashinc hash v = showHash $ (fctInc hash) $ v

makeTestAlg (name, hash, results) = testGroup name $ concatMap maketest (zip3 [0..] vectors results)
    where
        runtest :: ByteString -> String
        runtest v = runhash hash v

        runtestinc :: Int -> ByteString -> String
        runtestinc i v = runhashinc hash $ splitB i v

        maketest (i, v, r) =
            [ testCase (show i ++ " one-pass") (r @=? runtest v)
            , testCase (show i ++ " inc 1") (r @=? runtestinc 1 v)
            , testCase (show i ++ " inc 2") (r @=? runtestinc 2 v)
            , testCase (show i ++ " inc 3") (r @=? runtestinc 3 v)
            , testCase (show i ++ " inc 4") (r @=? runtestinc 4 v)
            , testCase (show i ++ " inc 5") (r @=? runtestinc 5 v)
            , testCase (show i ++ " inc 9") (r @=? runtestinc 9 v)
            , testCase (show i ++ " inc 16") (r @=? runtestinc 16 v)
            ]

katTests :: [TestTree]
katTests = map makeTestAlg results


main = defaultMain $ testGroup "cryptohash"
    [ testGroup "KATs" katTests
    ]


cbits/sha1.c
/*
 * Copyright (C) 2006-2009 Vincent Hanquez <vincent@snarc.org>
 *               2016      Herbert Valerio Riedel <hvr@gnu.org>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "sha1.h"

#include <assert.h>
#include <string.h>
#include <ghcautoconf.h>

#if defined(static_assert)
static_assert(offsetof(struct sha1_ctx, h[5]) == SHA1_CTX_SIZE, "unexpected sha1_ctx size");
#else
/* poor man's pre-C11 _Static_assert */
typedef char static_assertion__unexpected_sha1_ctx_size[(offsetof(struct sha1_ctx, h[5]) == SHA1_CTX_SIZE)?1:-1];
#endif

#define ptr_uint32_aligned(ptr) (!((uintptr_t)(ptr) & 0x3))

static inline uint32_t
rol32(const uint32_t word, const unsigned shift)
{
  /* GCC usually transforms this into a 'rol'-insn */
  return (word << shift) | (word >> (32 - shift));
}

static inline uint32_t
cpu_to_be32(const uint32_t hl)
{
#if WORDS_BIGENDIAN
  return hl;
#elif __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 2)
  return __builtin_bswap32(hl);
#else
  /* GCC usually transforms this into a bswap insn */
  return ((hl & 0xff000000) >> 24) |
         ((hl & 0x00ff0000) >> 8)  |
         ((hl & 0x0000ff00) << 8)  |
         ( hl               << 24);
#endif
}

static inline void
cpu_to_be32_array(uint32_t *dest, const uint32_t *src, unsigned wordcnt)
{
  while (wordcnt--)
    *dest++ = cpu_to_be32(*src++);
}

static inline uint64_t
cpu_to_be64(const uint64_t hll)
{
#if WORDS_BIGENDIAN
  return hll;
#elif __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 2)
  return __builtin_bswap64(hll);
#else
  return ((uint64_t)cpu_to_be32(hll & 0xffffffff) << 32LL) | cpu_to_be32(hll >> 32);
#endif
}


void
hs_cryptohash_sha1_init(struct sha1_ctx *ctx)
{
  memset(ctx, 0, SHA1_CTX_SIZE);

  ctx->h[0] = 0x67452301;
  ctx->h[1] = 0xefcdab89;
  ctx->h[2] = 0x98badcfe;
  ctx->h[3] = 0x10325476;
  ctx->h[4] = 0xc3d2e1f0;
}

#define f1(x, y, z)   (z ^ (x & (y ^ z)))
#define f2(x, y, z)   (x ^ y ^ z)
#define f3(x, y, z)   ((x & y) + (z & (x ^ y)))
#define f4(x, y, z)   f2(x, y, z)

#define K1  0x5a827999
#define K2  0x6ed9eba1
#define K3  0x8f1bbcdc
#define K4  0xca62c1d6

#define R(a, b, c, d, e, f, k, w)  \
        e += rol32(a, 5) + f(b, c, d) + k + w; b = rol32(b, 30)

#define M(i)  (w[i & 0x0f] = rol32(w[i & 0x0f] ^ w[(i - 14) & 0x0f] \
              ^ w[(i - 8) & 0x0f] ^ w[(i - 3) & 0x0f], 1))

static void
sha1_do_chunk_aligned(struct sha1_ctx *ctx, uint32_t w[])
{
        uint32_t a = ctx->h[0];
        uint32_t b = ctx->h[1];
        uint32_t c = ctx->h[2];
        uint32_t d = ctx->h[3];
        uint32_t e = ctx->h[4];

        R(a, b, c, d, e, f1, K1, w[0]);
        R(e, a, b, c, d, f1, K1, w[1]);
        R(d, e, a, b, c, f1, K1, w[2]);
        R(c, d, e, a, b, f1, K1, w[3]);
        R(b, c, d, e, a, f1, K1, w[4]);
        R(a, b, c, d, e, f1, K1, w[5]);
        R(e, a, b, c, d, f1, K1, w[6]);
        R(d, e, a, b, c, f1, K1, w[7]);
        R(c, d, e, a, b, f1, K1, w[8]);
        R(b, c, d, e, a, f1, K1, w[9]);
        R(a, b, c, d, e, f1, K1, w[10]);
        R(e, a, b, c, d, f1, K1, w[11]);
        R(d, e, a, b, c, f1, K1, w[12]);
        R(c, d, e, a, b, f1, K1, w[13]);
        R(b, c, d, e, a, f1, K1, w[14]);
        R(a, b, c, d, e, f1, K1, w[15]);
        R(e, a, b, c, d, f1, K1, M(16));
        R(d, e, a, b, c, f1, K1, M(17));
        R(c, d, e, a, b, f1, K1, M(18));
        R(b, c, d, e, a, f1, K1, M(19));

        R(a, b, c, d, e, f2, K2, M(20));
        R(e, a, b, c, d, f2, K2, M(21));
        R(d, e, a, b, c, f2, K2, M(22));
        R(c, d, e, a, b, f2, K2, M(23));
        R(b, c, d, e, a, f2, K2, M(24));
        R(a, b, c, d, e, f2, K2, M(25));
        R(e, a, b, c, d, f2, K2, M(26));
        R(d, e, a, b, c, f2, K2, M(27));
        R(c, d, e, a, b, f2, K2, M(28));
        R(b, c, d, e, a, f2, K2, M(29));
        R(a, b, c, d, e, f2, K2, M(30));
        R(e, a, b, c, d, f2, K2, M(31));
        R(d, e, a, b, c, f2, K2, M(32));
        R(c, d, e, a, b, f2, K2, M(33));
        R(b, c, d, e, a, f2, K2, M(34));
        R(a, b, c, d, e, f2, K2, M(35));
        R(e, a, b, c, d, f2, K2, M(36));
        R(d, e, a, b, c, f2, K2, M(37));
        R(c, d, e, a, b, f2, K2, M(38));
        R(b, c, d, e, a, f2, K2, M(39));

        R(a, b, c, d, e, f3, K3, M(40));
        R(e, a, b, c, d, f3, K3, M(41));
        R(d, e, a, b, c, f3, K3, M(42));
        R(c, d, e, a, b, f3, K3, M(43));
        R(b, c, d, e, a, f3, K3, M(44));
        R(a, b, c, d, e, f3, K3, M(45));
        R(e, a, b, c, d, f3, K3, M(46));
        R(d, e, a, b, c, f3, K3, M(47));
        R(c, d, e, a, b, f3, K3, M(48));
        R(b, c, d, e, a, f3, K3, M(49));
        R(a, b, c, d, e, f3, K3, M(50));
        R(e, a, b, c, d, f3, K3, M(51));
        R(d, e, a, b, c, f3, K3, M(52));
        R(c, d, e, a, b, f3, K3, M(53));
        R(b, c, d, e, a, f3, K3, M(54));
        R(a, b, c, d, e, f3, K3, M(55));
        R(e, a, b, c, d, f3, K3, M(56));
        R(d, e, a, b, c, f3, K3, M(57));
        R(c, d, e, a, b, f3, K3, M(58));
        R(b, c, d, e, a, f3, K3, M(59));

        R(a, b, c, d, e, f4, K4, M(60));
        R(e, a, b, c, d, f4, K4, M(61));
        R(d, e, a, b, c, f4, K4, M(62));
        R(c, d, e, a, b, f4, K4, M(63));
        R(b, c, d, e, a, f4, K4, M(64));
        R(a, b, c, d, e, f4, K4, M(65));
        R(e, a, b, c, d, f4, K4, M(66));
        R(d, e, a, b, c, f4, K4, M(67));
        R(c, d, e, a, b, f4, K4, M(68));
        R(b, c, d, e, a, f4, K4, M(69));
        R(a, b, c, d, e, f4, K4, M(70));
        R(e, a, b, c, d, f4, K4, M(71));
        R(d, e, a, b, c, f4, K4, M(72));
        R(c, d, e, a, b, f4, K4, M(73));
        R(b, c, d, e, a, f4, K4, M(74));
        R(a, b, c, d, e, f4, K4, M(75));
        R(e, a, b, c, d, f4, K4, M(76));
        R(d, e, a, b, c, f4, K4, M(77));
        R(c, d, e, a, b, f4, K4, M(78));
        R(b, c, d, e, a, f4, K4, M(79));

        ctx->h[0] += a;
        ctx->h[1] += b;
        ctx->h[2] += c;
        ctx->h[3] += d;
        ctx->h[4] += e;
}

static void
sha1_do_chunk(struct sha1_ctx *ctx, const uint8_t buf[])
{
  uint32_t w[16];
  if (ptr_uint32_aligned(buf)) { /* aligned buf */
    cpu_to_be32_array(w, (const uint32_t *)buf, 16);
  } else { /* unaligned buf */
    memcpy(w, buf, 64);
#if !WORDS_BIGENDIAN
    cpu_to_be32_array(w, w, 16);
#endif
  }
  sha1_do_chunk_aligned(ctx, w);
}

void
hs_cryptohash_sha1_update(struct sha1_ctx *ctx, const uint8_t *data, size_t len)
{
  size_t index = ctx->sz & 0x3f;
  const size_t to_fill = 64 - index;

  ctx->sz += len;

  /* process partial buffer if there's enough data to make a block */
  if (index && len >= to_fill) {
    memcpy(ctx->buf + index, data, to_fill);
    sha1_do_chunk(ctx, ctx->buf);
    /* memset(ctx->buf, 0, 64); */
    len -= to_fill;
    data += to_fill;
    index = 0;
  }

  /* process as many 64-blocks as possible */
  while (len >= 64) {
    sha1_do_chunk(ctx, data);
    len -= 64;
    data += 64;
  }

  /* append data into buf */
  if (len)
    memcpy(ctx->buf + index, data, len);
}

void
hs_cryptohash_sha1_finalize(struct sha1_ctx *ctx, uint8_t *out)
{
  static const uint8_t padding[64] = { 0x80, };

  /* add padding and update data with it */
  uint64_t bits = cpu_to_be64(ctx->sz << 3);

  /* pad out to 56 */
  const size_t index = ctx->sz & 0x3f;
  const size_t padlen = (index < 56) ? (56 - index) : ((64 + 56) - index);
  hs_cryptohash_sha1_update(ctx, padding, padlen);

  /* append length */
  hs_cryptohash_sha1_update(ctx, (uint8_t *) &bits, sizeof(bits));

  /* output hash */
  cpu_to_be32_array((uint32_t *) out, ctx->h, 5);
}


cbits/sha1.h
/*
 * Copyright (C) 2006-2009 Vincent Hanquez <vincent@snarc.org>
 *               2016      Herbert Valerio Riedel <hvr@gnu.org>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef CRYPTOHASH_SHA1_H
#define CRYPTOHASH_SHA1_H

#include <stdint.h>

struct sha1_ctx
{
	uint64_t sz;
	uint8_t  buf[64];
	uint32_t h[5];
};

#define SHA1_DIGEST_SIZE	20
#define SHA1_CTX_SIZE 		(sizeof(struct sha1_ctx))

void hs_cryptohash_sha1_init(struct sha1_ctx *ctx);
void hs_cryptohash_sha1_update(struct sha1_ctx *ctx, uint8_t *data, uint32_t len);
void hs_cryptohash_sha1_finalize(struct sha1_ctx *ctx, uint8_t *out);

#endif


Bench/Bench.hs
{-# LANGUAGE BangPatterns #-}
import Criterion.Main
import qualified Data.ByteString as B
import qualified Data.ByteString.Lazy as L
import qualified Crypto.Hash.SHA1 as SHA1

hashmany (i,u,f) = f . foldl u i

allHashs =
    [ ("SHA-1",SHA1.hash, hashmany (SHA1.init,SHA1.update,SHA1.finalize))
    ]

benchHash :: a -> (a -> B.ByteString) -> Benchmarkable
benchHash bs f = whnf f bs

withHashesFilter out f = map f $ filter (\(n,_,_) -> not (n `elem` out)) allHashs
withHashes f = map f allHashs

main = do
    let !bs32     = B.replicate 32 0
        !bs256    = B.replicate 256 0
        !bs4096   = B.replicate 4096 0
        !bs1M     = B.replicate (1*1024*1024) 0
    let !lbs64x256 = (map (const (B.replicate 64 0)) [0..3])
        !lbs64x4096 = (map (const (B.replicate 64 0)) [0..63])
    defaultMain
        [ bgroup "hash-32b" (withHashes (\(name, f,_) -> bench name $ benchHash bs32 f))
        , bgroup "hash-256b" (withHashes (\(name, f,_) -> bench name $ benchHash bs256 f))
        , bgroup "hash-4Kb" (withHashes (\(name, f,_) -> bench name $ benchHash bs4096 f))
        , bgroup "hash-1Mb" (withHashesFilter ["MD2"] (\(name, f,_) -> bench name $ benchHash bs1M f))
        , bgroup "iuf-64x256" (withHashes (\(name, _,f) -> bench name $ benchHash lbs64x256 f))
        , bgroup "iuf-64x4096" (withHashes (\(name, _,f) -> bench name $ benchHash lbs64x4096 f))
        ]


Bench/BenchAPI.hs
{-# LANGUAGE BangPatterns #-}
import Criterion.Main
import qualified Data.ByteString as B
import qualified Data.ByteString.Lazy as L
import qualified Crypto.Hash.SHA1 as SHA1

sha1F = ( "SHA1"
        , SHA1.hash
        , SHA1.finalize . SHA1.update SHA1.init
        )

main = do
    let !bs32     = B.replicate 32 0
        !bs256    = B.replicate 256 0
        !bs4096   = B.replicate 4096 0
        !bs1M     = B.replicate (1*1024*1024) 0
    let !lbs64x256 = (map (const (B.replicate 64 0)) [0..3])
        !lbs64x4096 = (map (const (B.replicate 64 0)) [0..63])

    let (fname, fHash, fIncr) = sha1F
    let benchName ty z = fname ++ "." ++ ty -- ++ " " ++ show z
    defaultMain
        [ bgroup "hash-0b"
            [ bench (benchName "hash" 0) $ whnf fHash B.empty
            , bench (benchName "incr" 0) $ whnf fIncr B.empty
            ]
        , bgroup "hash-32b"
            [ bench (benchName "hash" 32) $ whnf fHash bs32
            , bench (benchName "incr" 32) $ whnf fIncr bs32
            ]
        , bgroup "hash-256b"
            [ bench (benchName "hash" 256) $ whnf fHash bs256
            , bench (benchName "incr" 256) $ whnf fIncr bs256
            ]
        , bgroup "hash-4Kb"
            [ bench (benchName "hash" 4096) $ whnf fHash bs4096
            , bench (benchName "incr" 4096) $ whnf fIncr bs4096
            ]
        ]


src-bench/bench-sha1.hs
{-# LANGUAGE BangPatterns #-}

import           Criterion.Main
import qualified Crypto.Hash.SHA1     as SHA1
import qualified Data.ByteString      as B
import qualified Data.ByteString.Lazy as L

benchSize :: Int -> Benchmark
benchSize sz = bs `seq` bench msg (whnf SHA1.hash bs)
  where
    bs = B.replicate sz 0
    msg = "bs-" ++ show sz

main :: IO ()
main = do
    let !lbs64x256  = L.fromChunks $ replicate 4  (B.replicate 64 0)
        !lbs64x4096 = L.fromChunks $ replicate 64 (B.replicate 64 0)
    defaultMain
        [ bgroup "cryptohash-sha1"
          [ benchSize 0
          , benchSize 8
          , benchSize 32
          , benchSize 64
          , benchSize 128
          , benchSize 256
          , benchSize 1024
          , benchSize 4096
          , benchSize 8192
          , benchSize 16384
          , benchSize (128*1024)
          , benchSize (1024*1024)
          , benchSize (2*1024*1024)
          , benchSize (4*1024*1024)

          , L.length lbs64x256  `seq` bench "lbs64x256"  (whnf SHA1.hashlazy lbs64x256)
          , L.length lbs64x4096 `seq` bench "lbs64x4096" (whnf SHA1.hashlazy lbs64x4096)
          ]
        ]


src-tests/test-sha1.hs
{-# LANGUAGE OverloadedStrings #-}

module Main (main) where

import           Data.ByteString        (ByteString)
import qualified Data.ByteString        as B
import qualified Data.ByteString.Lazy   as BL
import qualified Data.ByteString.Base16 as B16

-- reference implementation
import qualified Data.Digest.Pure.SHA   as REF

-- implementation under test
import qualified Crypto.Hash.SHA1       as IUT

import           Test.Tasty
import           Test.Tasty.HUnit
import           Test.Tasty.QuickCheck  as QC

vectors :: [ByteString]
vectors =
    [ ""
    , "The quick brown fox jumps over the lazy dog"
    , "The quick brown fox jumps over the lazy cog"
    , "abc"
    , "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
    , "abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmnoijklmnopjklmnopqklmnopqrlmnopqrsmnopqrstnopqrstu"
    , B.replicate 1000000 0x61
    ]

answers :: [ByteString]
answers = map (B.filter (/= 0x20))
    [ "da39a3ee 5e6b4b0d 3255bfef 95601890 afd80709"
    , "2fd4e1c6 7a2d28fc ed849ee1 bb76e739 1b93eb12"
    , "de9f2c7f d25e1b3a fad3e85a 0bd17d9b 100db4b3"
    , "a9993e36 4706816a ba3e2571 7850c26c 9cd0d89d"
    , "84983e44 1c3bd26e baae4aa1 f95129e5 e54670f1"
    , "a49b2446 a02c645b f419f995 b6709125 3a04a259"
    , "34aa973c d4c4daa4 f61eeb2b dbad2731 6534016f"
    ]

ansXLTest :: ByteString
ansXLTest = B.filter (/= 0x20)
    "7789f0c9 ef7bfc40 d9331114 3dfbe69e 2017f592"

katTests :: [TestTree]
katTests
  | length vectors == length answers = map makeTest (zip3 [1::Int ..] vectors answers) ++ [xltest, xltest']
  | otherwise = error "vectors/answers length mismatch"
  where
    makeTest (i, v, r) = testGroup ("vec"++show i) $
        [ testCase "one-pass" (r @=? runTest v)
        , testCase "one-pass'" (r @=? runTest' v)
        , testCase "inc-1"    (r @=? runTestInc 1 v)
        , testCase "inc-2"    (r @=? runTestInc 2 v)
        , testCase "inc-3"    (r @=? runTestInc 3 v)
        , testCase "inc-4"    (r @=? runTestInc 4 v)
        , testCase "inc-5"    (r @=? runTestInc 5 v)
        , testCase "inc-7"    (r @=? runTestInc 7 v)
        , testCase "inc-8"    (r @=? runTestInc 8 v)
        , testCase "inc-9"    (r @=? runTestInc 9 v)
        , testCase "inc-16"   (r @=? runTestInc 16 v)
        , testCase "lazy-1"   (r @=? runTestLazy 1 v)
        , testCase "lazy-2"   (r @=? runTestLazy 2 v)
        , testCase "lazy-7"   (r @=? runTestLazy 7 v)
        , testCase "lazy-8"   (r @=? runTestLazy 8 v)
        , testCase "lazy-16"  (r @=? runTestLazy 16 v)
        , testCase "lazy-1'"   (r @=? runTestLazy' 1 v)
        , testCase "lazy-2'"   (r @=? runTestLazy' 2 v)
        , testCase "lazy-7'"   (r @=? runTestLazy' 7 v)
        , testCase "lazy-8'"   (r @=? runTestLazy' 8 v)
        , testCase "lazy-16'"  (r @=? runTestLazy' 16 v)
        ] ++
        [ testCase "lazy-63u"  (r @=? runTestLazyU 63 v) | B.length v > 63 ] ++
        [ testCase "lazy-65u"  (r @=? runTestLazyU 65 v) | B.length v > 65 ] ++
        [ testCase "lazy-97u"  (r @=? runTestLazyU 97 v) | B.length v > 97 ] ++
        [ testCase "lazy-131u" (r @=? runTestLazyU 131 v) | B.length v > 131] ++
        [ testCase "lazy-63u'"  (r @=? runTestLazyU' 63 v) | B.length v > 63 ] ++
        [ testCase "lazy-65u'"  (r @=? runTestLazyU' 65 v) | B.length v > 65 ] ++
        [ testCase "lazy-97u'"  (r @=? runTestLazyU' 97 v) | B.length v > 97 ] ++
        [ testCase "lazy-131u'" (r @=? runTestLazyU' 131 v) | B.length v > 131 ]
        
    runTest :: ByteString -> ByteString
    runTest = B16.encode . IUT.hash

    runTest' :: ByteString -> ByteString
    runTest' = B16.encode . IUT.finalize . IUT.start

    runTestInc :: Int -> ByteString -> ByteString
    runTestInc i = B16.encode . IUT.finalize . myfoldl' IUT.update IUT.init . splitB i

    runTestLazy :: Int -> ByteString -> ByteString
    runTestLazy i = B16.encode . IUT.hashlazy . BL.fromChunks . splitB i

    runTestLazy' :: Int -> ByteString -> ByteString
    runTestLazy' i = B16.encode . IUT.finalize . IUT.startlazy . BL.fromChunks . splitB i

    -- force unaligned md5-blocks
    runTestLazyU :: Int -> ByteString -> ByteString
    runTestLazyU i = B16.encode . IUT.hashlazy . BL.fromChunks . map B.copy . splitB i

    runTestLazyU' :: Int -> ByteString -> ByteString
    runTestLazyU' i = B16.encode . IUT.finalize . IUT.startlazy . BL.fromChunks . map B.copy . splitB i

    ----

    xltest = testGroup "XL-vec"
        [ testCase "inc" (ansXLTest @=? (B16.encode . IUT.hashlazy) vecXL) ]
      where
        vecXL = BL.fromChunks (replicate 16777216 "abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmno")

    xltest' = testGroup "XL-vec'"
        [ testCase "inc'" (ansXLTest @=? (B16.encode . IUT.finalize . IUT.startlazy) vecXL) ]
      where
        vecXL = BL.fromChunks (replicate 16777216 "abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmno")


splitB :: Int -> ByteString -> [ByteString]
splitB l b
  | B.length b > l = b1 : splitB l b2
  | otherwise = [b]
  where
    (b1, b2) = B.splitAt l b


rfc2202Vectors :: [(ByteString,ByteString,ByteString)]
rfc2202Vectors = -- (secrect,msg,mac)
    [ (rep 20 0x0b, "Hi There", x"b617318655057264e28bc0b6fb378c8ef146be00")
    , ("Jefe", "what do ya want for nothing?", x"effcdf6ae5eb2fa2d27416d5f184df9c259a7c79")
    , (rep 20 0xaa, rep 50 0xdd, x"125d7342b9ac11cd91a39af48aa17b4f63f175d3")
    , (B.pack [1..25], rep 50 0xcd, x"4c9007f4026250c6bc8414f9bf50c86c2d7235da")
    , (rep 20 0x0c, "Test With Truncation", x"4c1a03424b55e07fe7f27be1d58bb9324a9a5a04")
    , (rep 80 0xaa, "Test Using Larger Than Block-Size Key - Hash Key First", x"aa4ae5e15272d00e95705637ce8a3b55ed402112")
    , (rep 80 0xaa, "Test Using Larger Than Block-Size Key and Larger Than One Block-Size Data", x"e8e99d0f45237d786d6bbaa7965c7808bbff1a91")
    , (rep 80 0xaa, "Test Using Larger Than Block-Size Key - Hash Key First", x"aa4ae5e15272d00e95705637ce8a3b55ed402112")
    , (rep 80 0xaa, "Test Using Larger Than Block-Size Key and Larger Than One Block-Size Data", x"e8e99d0f45237d786d6bbaa7965c7808bbff1a91")
    ]
  where
    x = B16.decodeLenient
    rep n c = B.replicate n c

rfc2202Tests = zipWith makeTest [1::Int ..] rfc2202Vectors
  where
    makeTest i (key, msg, mac) = testGroup ("vec"++show i) $
        [ testCase "hmac" (hex mac  @=? hex (IUT.hmac key msg))
        , testCase "hmaclazy" (hex mac  @=? hex (IUT.hmaclazy key lazymsg))
        ]
      where
        lazymsg = BL.fromChunks . splitB 1 $ msg

    hex = B16.encode

-- define own 'foldl' here to avoid RULE rewriting to 'hashlazy'
myfoldl' :: (b -> a -> b) -> b -> [a] -> b
myfoldl' f z0 xs0 = lgo z0 xs0
  where
    lgo z []     = z
    lgo z (x:xs) = let z' = f z x
                   in z' `seq` lgo z' xs

newtype RandBS = RandBS { unRandBS :: ByteString }
newtype RandLBS = RandLBS BL.ByteString

instance Arbitrary RandBS where
    arbitrary = fmap (RandBS . B.pack) arbitrary
    shrink (RandBS x) = fmap RandBS (go x)
      where
        go bs = zipWith B.append (B.inits bs) (tail $ B.tails bs)

instance Show RandBS where
    show (RandBS x) = "RandBS {len=" ++ show (B.length x)++"}"

instance Arbitrary RandLBS where
    arbitrary = fmap (RandLBS . BL.fromChunks . map unRandBS) arbitrary

instance Show RandLBS where
    show (RandLBS x) = "RandLBS {len=" ++ show (BL.length x) ++ ", chunks=" ++ show (length $ BL.toChunks x)++"}"


refImplTests :: [TestTree]
refImplTests =
    [ testProperty "hash" prop_hash
    , testProperty "start" prop_hash
    , testProperty "hashlazy" prop_hashlazy
    , testProperty "startlazy" prop_startlazy
    , testProperty "hmac" prop_hmac
    , testProperty "hmaclazy" prop_hmaclazy
    ]
  where
    prop_hash (RandBS bs)
        = ref_hash bs == IUT.hash bs

    prop_start (RandBS bs)
        = ref_hash bs == (IUT.finalize $ IUT.start bs)

    prop_hashlazy (RandLBS bs)
        = ref_hashlazy bs == IUT.hashlazy bs

    prop_startlazy (RandLBS bs)
        = ref_hashlazy bs == (IUT.finalize $ IUT.startlazy bs)

    prop_hmac (RandBS k) (RandBS bs)
        = ref_hmac k bs == IUT.hmac k bs

    prop_hmaclazy (RandBS k) (RandLBS bs)
        = ref_hmaclazy k bs == IUT.hmaclazy k bs

    ref_hash :: ByteString -> ByteString
    ref_hash = ref_hashlazy . fromStrict

    ref_hashlazy :: BL.ByteString -> ByteString
    ref_hashlazy = toStrict . REF.bytestringDigest . REF.sha1

    ref_hmac :: ByteString -> ByteString -> ByteString
    ref_hmac secret = ref_hmaclazy secret . fromStrict

    ref_hmaclazy :: ByteString -> BL.ByteString -> ByteString
    ref_hmaclazy secret = toStrict . REF.bytestringDigest . REF.hmacSha1 (fromStrict secret)

    -- toStrict/fromStrict only available with bytestring-0.10 and later
    toStrict = B.concat . BL.toChunks
    fromStrict = BL.fromChunks . (:[])

main :: IO ()
main = defaultMain $ testGroup "cryptohash-sha1"
    [ testGroup "KATs" katTests
    , testGroup "RFC2202" rfc2202Tests
    , testGroup "REF" refImplTests
    ]
